### ISSUE TITLE
Does Geany Preview exist for Lubuntu 23.10?

### ISSUE THREAD

**Issue by ralf3u** [xiota/geany-preview#49]
I ask this because I would like to test Lubuntu 23.10.
Thank you for an answer.

**Comment by xiota** [comment]
I will try to update the PPA this weekend.

Geany 2.x was recently released.  Do you want 2.x added to the PPA?  I think it includes fixes for some bugs that were affecting you.  As far as I can tell, 2.x is not yet in official Ubuntu repositories.  I wouldn't expect it until the 24.04 release.

The problem is, plugins built for 1.38 won't work on 2.x anymore, and vice versa.

**Comment by ralf3u** [comment]
>Geany 2.x was recently released. Do you want 2.x added to the PPA?

No, thanks. Geany Preview is enough.

**Comment by xiota** [comment]
The latest version of the plugin has been built for 22.04, 23.04, and 23.10 on the PPA.

The plugin no longer builds on 20.04 because of API changes.  An older version may still be available.

I deleted the old build of geany-git.  I'm leaving geany-plugins because Lua and some other plugins had some bug fixes.

**Comment by ralf3u** [comment]
> I'm leaving geany-plugins [...]

Oh. What does "leaving" in concrete mean?

>[...] because Lua and some other plugins had some bug fixes.

Does that mean that the shortcut to disable/enable the menu bar of Geany does not work any more?
Does that mean that the auto-sidebar and the shortcuts for Geany Preview don't work any more?

**Comment by xiota** [comment]
> What does "leaving" in concrete mean? ... does not work any more? ...

It means, I am not deleting them.  However, I had forgotten that they need to be built to be available for 23.10.  I will do so in the next few days.  Sorry for the oversight.

**Comment by ralf3u** [comment]
>It means, I am not deleting them. However, I had forgotten that they need to be built to be available for 23.10.

I don't understand. If I install geany and geany-plugins, then Lua Scripts should be there. So, why Lua Scripts need to be built?

**Comment by xiota** [comment]
I recall that there were bugs in the Lua scripts plugin that are not fixed in the version currently provided by Debian/Ubuntu.  If that is not the case, then I do not need to rebuild the plugins for the PPA.  I will check the package database before building.

**Comment by ralf3u** [comment]
>[...] there were bugs in the Lua scripts plugin that are not fixed in the version currently provided by Debian/Ubuntu.

Does this also have consequences on the  shortcut to disable/enable the menu bar of Geany, the auto-sidebar and the shortcuts for Geany Preview? If not, then I don't care so much if there are bugs in Lua scripts.

**Comment by ralf3u** [comment]
Sorry to answer so late. But I was confronted with many problems. I just wanted to use the toggle menu bar and test later Preview. But I can't use the toggle menu bar because I can't find the plugin 'Lua Script' in the list of plugins.

**Comment by ralf3u** [comment]
If you are too busy at the moment, I can wait till end of April, because in two months there will be Lubuntu 24.04.

**Comment by xiota** [comment]
I haven't been using Ubuntu-based distros for a while...  Will review PPA over the few days or weekend.

**Comment by ralf3u** [comment]
This year I would like to test again Debian, because in Debian there is the option to install only free software, what can be realized in the expert-mode. Some years ago I installed Debian, but I think that I was frustrated because the speakers didn't work. I thought that Debian was the problem. Today I think that the speakers didn't work because of the non-free drivers. What is your experience with Debian?

**Comment by xiota** [comment]
My info about Debian is old.  I used Debian for 1-2 years a long time ago.  I used Ubuntu variants for many years after that.

Debian package versions tend to be old relative to the latest Ubuntu releases, but not too much different from using LTS.  Debian default config is different.  Package availability is different.  PPA may not work with Debian.

In terms of free vs non-free, most firmware is considered non-free.  Without firmware, many devices do not work at all.  So usually hardware issues are not fixed by removing non-free packages.  (Different issue from proprietary vs open-source drivers.)  But only way to really know is to try.

**Comment by ralf3u** [comment]
>PPA may not work with Debian.

Hm, so it could be that Geany Preview does not work in Debian?

**Comment by xiota** [comment]
The plugin should still be usable.  You'd have you build it yourself though.  Standard debian packaging is a pain.  Something like [`makedeb`](https://github.com/makedeb/makedeb) and [MPR](https://mpr.makedeb.org) would be easier.

**Comment by xiota** [comment]
I had thought the packages were already built for 23.10 last Nov, but I had a typo in my config.  It should be fixed now.  Let me know if it's not working by tomorrow.  (You're too patient, waiting so many months.)

I'm thinking about retiring the PPA in favor of including a `makedeb` config file.

**Comment by xiota** [comment]
> I still can't find Lua Scripts in the list of plugins in Geany.

Is `geany-plugin-lua` installed?

**Comment by ralf3u** [comment]
>Is geany-plugin-lua installed?

I can't remember that I had to install geany-plugin-lua in the past. It was automatically in the list of plugins after the installation of geany-plugins.

**Comment by ralf3u** [comment]
What is the reason that it is not any more integrated in geany-plugins?

**Comment by xiota** [comment]
I'd rather not add distro-specific instructions to install Lua.  Users should be able to figure out that much on their own.

**Comment by ralf3u** [comment]
>I'd rather not add distro-specific instructions to install Lua. Users should be able to figure out that much on their own.

OK. But if I install geany-plugin-lua and then if Lua Scripts is in the list of plugins, can I update the 'step-by-step description', that is located in the issue number 633 of Geany, by adding a kind of text like the next line?

`if Lua Scripts is not in the list of plugins, then you need to install Lua Scripts by terminal: sudo apt install geany-plugin-lua`

**Comment by xiota** [comment]
You're free to edit your own comment.  The reminder to check that the plugin is installed could be helpful, but not everyone uses Debian/Ubuntu, so the apt command won't necessarily help.  For example, if Geany is manually built and installed to `/usr/local` (outside package manager control), Geany plugins also has to be manually built and installed to `/usr/local`.

I wouldn't mind adding to the lua scripts readme:

> If "Lua Script" is not shown in the Plugin Manager, make sure the Geanylua plugin is installed.  See distro documentation for details.

**Comment by ralf3u** [comment]
I installed the plugin Lua Script with sudo apt install geany-plugin-lua.
Now, Lua Script is in the list of plugins.
I did exactly the 'step-by-step description' of issue number 633. But in the Preferences, in Keybindings, there is no Lua Script although it is activated in the list of plugins.

**Comment by ralf3u** [comment]
Sorry for my last comment. I found it now. I delete both comments.

**Comment by ralf3u** [comment]
I installed the plugin Lua Script with `sudo apt install geany-plugin-lua`.
Now, Lua Script is in the list of plugins.
The shortcut toggle menu bar works.
I updated the content of the `step-by-step description` at Geany (issue number 633).

In the next days I will install Geany Preview to test the other shortcuts.

**Comment by ralf3u** [comment]
Excuse me to disturb you, but I have a problem, and maybe you know the solution.

I can change the design of tabs and I can change the design of the active tabs, but I can't change the design of the tab, that is in focus. So, this is what works in geany.css:
```
/* design of tabs */
tab {
	padding-top:25px;
	padding-bottom:10px;
	padding-left: 35px;
}

/* design of active tabs (e.g. tab of the editor and tab of the sidebar) */
tab:checked {
	font-weight: bold;
	border-bottom: 5px solid black;
}
```
I opened yesterday a discussion at Geany with the number 3784, but the answer is not a help for me. I think you could know the solution, because when I used the shortcut 'Toggle between the editor and sidebar' in Preview, I could clearly see which tab is in focus.

So, do you know the selector name for geany.css to change the design of the tab that is in focus? Or is it not possible to solve the problem in geany.css? Is there an easy solution?

Thank you very much in advance for an answer.

**Comment by xiota** [comment]
What you want sounds like theming.  Maybe look at some gtk themes to see what they do.  May not be any universal solution.  Could be different for different themes.

**Comment by ralf3u** [comment]
>What you want sounds like theming. Maybe look at some gtk themes to see what they do. May not be any universal solution. Could be different for different themes.

I tested all GTK 3 Themes and some GTK 2 Themes in main menu, Preferences, LXQt Settings, Appearance, GTK Style. After each change I restarted Geany. But it didn't help.

For the sidebar I just recognized that you already programmed a solution for sidebar-tab-focus/unfocus. I described how to use it in https://github.com/xiota/geany-preview/issues/5#issuecomment-969031972. So, the focus/unfocus problem of the sidebar could be solved partially with `geany-plugin-xi-tweaks`,  partially because #5 still exists.

Thanks to geany-plugin-xi-tweaks, now, what I need is a solution to recognize the tab-unfocus of the editor. It could be like this:
```
tab:checked {
	font-weight: bold;
	border-bottom: 5px solid black;
}
tab:unfocus {
	font-weight: bold;
}
```

But `tab:unfocus` does not work. 

In https://github.com/xiota/geany-preview/issues/5#issuecomment-939193808 you write: "The editor is a different component. (scintilla)".

Do you have an idea how to recognize the tab-unfocus of the editor?

Thank you in advance for an answer.

**Comment by ralf3u** [comment]
The problem has to do with Preview. If I use the default settings of geany.css, then I can see which tab is in focus when pressing the shortcut `Switch focus between editor and the sidebar`. But because I use an ePaper-display, I can't use the default settings of geany.css. So, I use tab:checked. But when pressing the shortcut `Switch focus between editor and the sidebar` I would like to use a kind of  tab:unchecked or rather tab:unfocus or tab:unactive for the tab in the editor.

**Comment by xiota** [comment]
Maybe some screenshots or photos would help me understand better what you are describing?

I'm also interested in the epaper display.  What are the make and model?

**Comment by ralf3u** [comment]
>I'm also interested in the epaper display. What are the make and model?

It is BOOX MIRA (boox.com). It is so good for the eyes. There is also a BOOX MIRA PRO, but it is more for music production or video production because of the big screen. I had to change to ePaper-display because the eyes started to hurt after a while when using the notebook display.  When using a TFT/LCD-display, you look into light. And that is not good for the eyes on a long term. In BOOX MIRA I can turn off the lights, so I don't see into the light. I use a LED-lamp instead with 11 Watt that shines directly into the display, so I don't see into the light, but into reflected light, what is much better for the eyes.

**Comment by ralf3u** [comment]
>Maybe some screenshots or photos would help me understand better what you are describing?

Just paste the following code at the end of geany.css, save geany.css, and then restart Geany by opening the sidebar.
```
/* design of tabs */
tab {
	padding-top:25px;
	padding-bottom:10px;
	padding-left: 35px;
}

/* design of active tabs (e.g. tab of the editor and tab of the sidebar) */
tab:checked {
	font-weight: bold;
	border-bottom: 5px solid black;
}
````
Then you will see that both active tabs have exactly the same design. And then, if you put a tab of the sidebar in focus, the design of the tab of the editor does not change.

**Comment by xiota** [comment]
I think I see what you're trying to do now.  You want whichever single tab is active / focused / receiving keyboard input, regardless of which notebook it is in (editor or sidebar) to have a visual indicator to show its state.

I don't know how.  If possible, you probably need to find a GTK-specific forum.  `:focus` seems to be closest, but it doesn't work.  Maybe because the tab children have focus  (eg, the editor), not the tab itself.

The xi-tweaks solution is hacky and I recall it didn't work in some cases, like when you changed tabs using the keyboard.

**Comment by xiota** [comment]
I'm modifying xi-tweaks to highlight a single active tab across editor, sidebar, and msgwin.  It's basically working with mouse clicks, but not when the tab is changed using the keyboard.  It's also freezing on shutdown.  If I figure out why it's freezing, I'll upload it to the PPA for you to test.

**Comment by xiota** [comment]
I've updated the xi-tweaks plugin to highlight a single active tab across editor, sidebar, and msgwin.  It's uploaded to the PPA: `geany-plugin-xi-tweaks 0.0.0-r32.g9900a17`

Config:
```
[tweaks]
sidebar_focus_enabled=false
notebook_focus_enabled=true
```

`sidebar_focus_enabled=true` is the old behavior that highlights only the sidebar tabs.

`notebook_focus_enabled=true` highlights the single focused tab from sidebar, msgwin, or editor.  "Notebook" is the GTK terminology for a tabbed widget.

`geany.css` rules are different from before.  Here is a sample config:
```
#geany-xitweaks-notebook-tab-focus {
	font-weight: bold;
	color: #399;
}

#geany-xitweaks-notebook-tab-unfocus {
}

#geany-xitweaks-notebook-page-focus {
	border-width: 3px;
	border-style: solid;
	border-color: #399;
}

#geany-xitweaks-notebook-page-unfocus {
	border-width: 3px;
	border-style: solid;
	border-color: transparent;
}
```

Basically, `sidebar` in old name is changed to `notebook`.  The example config installed with the package has an unused example I forgot to remove.

**Comment by ralf3u** [comment]
I have three questions:

In my description of https://github.com/xiota/geany-preview/issues/5#issuecomment-969031972 I write that geany-plugin-xi-tweaks can be installed by the following command:
`sudo apt install geany-plugin-xi-tweaks`

In your last comment above you write:
>It's uploaded to the PPA:
geany-plugin-xi-tweaks 0.0.0-r32.g9900a17

1) Does that mean that I have first to install Preview?
2) If the answer is yes, can other users still use the following command?
`sudo apt install geany-plugin-xi-tweaks`
3) What does msgwin mean?

**Comment by xiota** [comment]
The xi-tweaks plugin is separate from the preview plugin.

If the PPA is installed, the command will install the xi-tweaks plugin should be `sudo apt install geany-plugin-xi-tweaks`

The css rules related to `#geany-preview-sidebar-page` are not used by the xi-tweaks or preview plugins.  It is left over from some previous attempts to provide the tab highlighting feature you requested.

The other css rules have been renamed by changing `sidebar` to `notebook`.  For example `#geany-xitweaks-sidebar-page-unfocus` is now `#geany-xitweaks-notebook-page-unfocus`.

The message window (msgwin) is another tabbed panel.  You can see it by activating: Menubar > View > Show Message Window.

**Comment by ralf3u** [comment]
Just for information: During the installation process of the PPA  there was this text in the terminal:
```
Description:
20.04, 22.04, 22.10, 23.04
```
23.10 is missing

**Comment by xiota** [comment]
Thank you.  I updated the description to `20.04, 22.04, 23.10`.  `24.04` may be the last.

Let me know if xi-tweaks feature `notebook_focus_enabled=true` does what you were searching for.  Also, somehow the problem with wrong tab being highlighted with keyboard navigation may be fixed now.  No idea how/why.

**Comment by ralf3u** [comment]
With those settings, everything works correct:
```
#geany-xitweaks-notebook-tab-focus {
	font-weight: bold;
}

#geany-xitweaks-notebook-tab-unfocus {
}

#geany-xitweaks-notebook-page-focus {
}

#geany-xitweaks-notebook-page-unfocus {
}
```


But with those settings:
```
#geany-xitweaks-notebook-tab-focus {
	font-weight: bold;
	border-bottom: 5px solid black;
}

#geany-xitweaks-notebook-tab-unfocus {
}

#geany-xitweaks-notebook-page-focus {
}

#geany-xitweaks-notebook-page-unfocus {
}

```
If sidebar or msgwin are in focus: The tab-title of the sidebar or msgwin is shown with a border, as it should be.
But:
If editor is in focus: The tab-title of the sidebar is shown without a border, although I expect a border.

**Comment by xiota** [comment]
>  If editor is in focus: The tab-title of the sidebar is shown without a border, although I expect a border.

Maybe adding generic CSS, similar to your previous attempt, would do what you want?
```
tab:checked {
	border-bottom: 5px solid black;
}
```

**Comment by ralf3u** [comment]
```
tab:checked {
	border-bottom: 5px solid black;
}
```
Then all three tab-titles are underlined what is not my goal.

The goal for me:
tab that has focus: underline and bold; at the same time:
active tab of the other panels:  only bold

**Comment by xiota** [comment]
>  active tab of the other panels: only bold
```
tab:checked {
	font-weight: bold;
}
```

>  tab that has focus: underline and bold; at the same time:
```
#geany-xitweaks-notebook-tab-focus {
	font-weight: bold;
	border-bottom: 5px solid black;
}

#geany-xitweaks-notebook-tab-unfocus {
}

#geany-xitweaks-notebook-page-focus {
}

#geany-xitweaks-notebook-page-unfocus {
}
```

**Comment by ralf3u** [comment]
When the tab of the editor is in focus: There is no underline what is not the goal.

**Comment by xiota** [comment]
I don't understand what you are trying to accomplish then.  Maybe the following config?
```
[tweaks]
sidebar_focus_enabled=true
notebook_focus_enabled=false
```

**Comment by ralf3u** [comment]
>I don't understand what you are trying to accomplish then.

tab that has focus: underline and bold; at the same time:
active tab of the other panels: only bold

I will try your last comment in around ten hours.

**Comment by ralf3u** [comment]
First I want to say that you had a a good idea to add a further option to the plugin Xi/Tweaks and not to change completely the plugin, so that in case the further option does not work, there is still a fallback-solution.

I recognized that I still don't know exactly what kind of tab-title-format I would like to use because the frequent use will show what is the best for me.

Some days ago I wanted that only the tab that is in focus should be underlined and bold. But then I recognized that I lose the orientation on the ePaper-display because I don't know which tab is active of the panel that has not the focus. That is the reason why I chose that all active tabs should be bold and while the tab that is in focus should be underlined.

To make all active tabs bold can be realized by this:
```
tab:checked {
	font-weight: bold;
}
```
Then, the only thing what is missing is the underline.

To make the underline work not only in the sidebar but also in other panels I need this form of configuration:
```
sidebar_focus_enabled=false
notebook_focus_enabled=true
```
That is the reason why I decided today not to test your suggestion of your last comment.

Now comes the CSS for the underline. This is the part that is not working in the editor as it should work:
```
#geany-xitweaks-notebook-tab-focus {
	border-bottom: 5px solid black;
}

#geany-xitweaks-notebook-tab-unfocus {
}

#geany-xitweaks-notebook-page-focus {
}

#geany-xitweaks-notebook-page-unfocus {
}
```

But what is interesting: If I replace
`border-bottom: 5px solid black;`
by this line
`font-style:oblique;`
or by this line
`font-size: 50px;`
then it works in the editor.

So, the CSS-selector `border` is the problem, because if I chose this line
`border: 5px solid black;`
it also does not work in the editor while in the other panels it works.

Maybe the problem is related to the following lines because the border is touching the page?
```
#geany-xitweaks-notebook-page-focus {
}

#geany-xitweaks-notebook-page-unfocus {
}
```
It is weird. Do you you have an explication?

**Comment by ralf3u** [comment]
I think that my guess of my last comment is correct, because If I chose this
```
#geany-xitweaks-notebook-tab-focus {
}

#geany-xitweaks-notebook-tab-unfocus {
	border: 5px solid black;
}

#geany-xitweaks-notebook-page-focus {
	border: 5px solid black;
}

#geany-xitweaks-notebook-page-unfocus {
	border: 5px solid black;
}
```
I can clearly see that the page is above the area where the underline should appear.

Maybe the solution is to put the page of the editor under the tab bar.

**Comment by ralf3u** [comment]
The CSS-selector `text-decoration` does not work on any panel. I don't know why.

**Comment by xiota** [comment]
GTK CSS isn't the same has webpage CSS.  So not everything works the same.  Probably `text-decoration` is implemented differently or not at all.

Maybe you need a rule for a different name to do what you want.  Here's a [search for different names](https://github.com/search?q=repo%3Ageany%2Fgeany+gtk_widget_set_name&type=code) in Geany.

**Comment by ralf3u** [comment]
>Maybe you need a rule for a different name to do what you want. Here's a [search for different names](https://github.com/search?q=repo%3Ageany%2Fgeany+gtk_widget_set_name&type=code) in Geany.

I couldn't find something.

And what about the border problem?

**Comment by xiota** [comment]
The editor tab css seems to be forced somehow, limiting what you can do with it in `geany.css`.  If you want, you can ask geany devs, at one of the issues you mentioned earlier, why css rules aren't working on the editor tab and whether/how/where they're being overridden.

Maybe easier to put a border around the whole page:
```
#geany-xitweaks-notebook-page-focus {
	border-width: 5px;
	border-style: solid;
	border-color: black;
}

#geany-xitweaks-notebook-page-unfocus {
	border-width: 5px;
	border-style: solid;
	border-color: transparent;
}
```

**Comment by ralf3u** [comment]
The code of your last comment didn't help.

> If you want, you can ask geany devs [...]

Is the internet address down below correct to open issues at GTK?
https://gitlab.gnome.org/GNOME/gtk/-/issues

**Comment by xiota** [comment]
Try this:
```

#geany-xitweaks-notebook-tab-focus,
#geany-xitweaks-notebook-tab-focus label {
	font-weight: bold;

	border-bottom-width: 3px;
	border-style: solid;
	border-color: black;
}

#geany-xitweaks-notebook-tab-unfocus,
#geany-xitweaks-notebook-tab-unfocus label {
}
```

**Comment by ralf3u** [comment]
I tried your code from above and then the underline works in all panels!!!!

But the problem is then, that if one tab of the editor is in focus, and then if I click on one tab of the sidebar, then the content of the tab of the sidebar is moving downwards while the the content of the tab of the editor is moving upwards.

So I changed the code from your last comment, like this:
```
tab:checked {
	font-weight: bold;
}

#geany-xitweaks-notebook-tab-focus,
#geany-xitweaks-notebook-tab-focus label {
	border-bottom: 5px solid black;
}

#geany-xitweaks-notebook-tab-unfocus,
#geany-xitweaks-notebook-tab-unfocus label {
	border-bottom: 5px solid transparent;
}
```
And now the content of the tabs is not moving any more downwards/upwards.
It's so cool!!
Now it is exactly how I wanted.

Thank you so much for your work.

**Comment by ralf3u** [comment]
>The next Ubuntu release 24.04 may be the last.

It could be that I will stay with Lubuntu.
Will there be a possibility to install and use geany-plugin-lua, geany-plugin-xi-tweaks and Preview in Lubuntu 24.10?

**Comment by xiota** [comment]
I hope to migrate to a system for users to make packages from source. Something like [`makedeb`](https://www.makedeb.org/).

**Comment by ralf3u** [comment]
>I hope to migrate to a system for users to make packages from source. Something like makedeb.

I have not big knowledge about installations. Does that mean that  geany-plugin-lua, geany-plugin-xi-tweaks and Preview should work in Lubuntu 24.10, but the installations will be different?

**Comment by xiota** [comment]
The plan is for users to build the package on their own computers.  Then they would  install the package with `dpkg`.

The plugins build reasonably quickly, so it shouldn't be too inconvenient.

If you're still testing different distros, consider trying Manjaro, Endeavour, ALCI.

**Comment by ralf3u** [comment]
The plugin geany-plugin-xi-tweaks does not need to be changed any more because everything works.
Because the plugin geany-plugin-xi-tweaks is so good, what about to add it in the official list of plugins, so there is no need to install it by PPA or something else, like geany-plugin-lua?

In the past in the plugin geany-plugin-xi-tweaks there was a lot of things. Now it is only about tab-title of tabs. So, the name could be changed from geany-plugin-xi-tweaks to geany-plugin-tab-title.

**Comment by xiota** [comment]
Will think about renaming.  Since the features are working now, may add to my other tweaks plugin.

Would need approval to be added to geany-plugins.  Then would need approval for each future update.  Based on experience trying to fix bugs in other plugins, would rather not bother trying to make an official plugin.

**Comment by ralf3u** [comment]
>If you're still testing different distros, consider trying Manjaro, Endeavour, ALCI.

I'm only interested in Debian because of the possibility to install only free software.

I would like to write a step-by-step-description how to change the design of the tab-title of the tab that is in focus. I could do it in the discussions  with the number 3784 at geany, but I prefer to open a new issue in discussions in `Show and tell` (3784 is located in the area `Q&A`). I would like to post it at geany, because I can imagine that users have more trust in information that is located at geany. Or do you have an other idea?

**Comment by xiota** [comment]
What software to install depends on what the user chooses.  I suggested those distros because they're easier to make packages from source.

You should wait to write your howto until after the PPA and plugin reorganization.  Also, it's a niche feature.  Is there any indication any one else is interested?

**Comment by ralf3u** [comment]
>Also, it's a niche feature. Is there any indication any one else is interested?

Well, I have no idea.

>You should wait to write your howto until after the PPA and plugin reorganization.

OK.

**Comment by ralf3u** [comment]
Do you know if there is a possibility to change the font-size of the content area of the tab 'Documents' of the sidebar?

Reason:
The document names and the folder names in the content area of the tab 'Documents' of the sidebar are too small.

There is already an open issue at geany with the issue number 424.

Thank you in advance for an answer.

**Comment by xiota** [comment]
The only place I know is Preferences / Interface / Interface / Fonts.

**Comment by ralf3u** [comment]
Because you are an expert of Geany Tweaks I wanted to ask you if it is possible to add a 1px black border to the tab-switcher?
I opened yesterday at geany an issue with number 3797. The last paragraph of the last comment could show a way to the solution. But I don't know how to handle it. Do you have an idea?
Thank you in advance for a help.

**Comment by xiota** [comment]
As described in the referenced comment, used the [GTK Inspector](https://wiki.gnome.org/Projects/GTK/Inspector), by setting `gsettings set org.gtk.Settings.Debug enable-inspector-keybinding true` and pressing Control-Shift-D in Geany.

Found name `GeanyDialog`.  Try adding to `geany.css`:
```
#GeanyDialog {
	border-width: 2px;
	border-style: solid;
	border-color: black;
}
```

**Comment by ralf3u** [comment]
I added the code from the last comment in geany.css.
First in geany.css that is located in .config/geany/, and then in geany.css that is located in /usr/share/geany/.
After a restart of Geany the tab-switcher has still no border.

**Comment by xiota** [comment]
Please provide screenshot of what you are referring to as "tab-switcher".

**Comment by ralf3u** [comment]
>Please provide screenshot of what you are referring to as "tab-switcher".

At geany with the issue number 3800 there is thsi link:
https://imgur.com/mWzOafF
You can see the tab-switcher on the right side of the image.
The tab-switcher appears when pressing Shift+F10.

**Comment by xiota** [comment]
I was looking at the wrong popup.  I'm not sure the shift-F10 popup can be styled separately from other context menus without giving it a name in Geany.

**Comment by ralf3u** [comment]
I will test the switch-views in Lubuntu 24.04.

Thank you so much for Preview that works in Lubuntu 23.10.

**Comment by ralf3u** [comment]
I have two important questions outside Preview:
- Do you use speech-to-text-software?
- If yes, which one?

In Lubuntu 23.04 I used Vosk in combination with nerd-dictation. But in Lubuntu 23.10 it is not recommanded to install Vosk like this:
`pip3 install vosk`

It is recommanded to use those three commands during the installation: 
```
sudo apt install pipx
pipx ensurepath
pipx install vosk
```
Then vosk is running, but not nerd-dictation. I opened at github at nerd-dictation the issue with the number 119. But I get no answer.

In Lubuntu 23.04 I dictated right away in the editor. In Lubuntu 23.10 I have first to record the speech with Audacity, and then to transcribe the speech to text with Vosk.

Thank you in advance for an answer.

**Comment by xiota** [comment]
I haven't used dictation software.  It's an area I think Linux is lacking.  I'll look into the programs you mentioned.

I don't know difference between pip and pipx.  Why not use pip with venv?

**Comment by ralf3u** [comment]
>It's an area I think Linux is lacking.

There's no lack in this area. There's a lack of information that the speech-to-text-recognition software is so good. I used it in Lubuntu 23.04 and I was able to dictate text in Geany and LibreOffice. That was really awesome. The recognition was really good. I used three shortcuts for starting the program, one for French, one for German and one for English. 16 GB RAM are necessary to run Vosk. But for me it worked also by using 8 GB RAM and 8 GB swap-file to have in sum 16 GB. Even with real 16 GB RAM it took many seconds so that Vosk is ready for dictation (up to 40 seconds I think so). Vosk does not understand punctuation and design, like "dot" or "comma" or "colon" or "go to the next line". The dots and commas can be set later with an other program and I was astonished how good the result is. The punctuation was perfect.

>Why not use pip with venv?

I have no idea what pip and pix and python and venv means. I just do the steps. I just recognized that someone else wrote right after me an issue at nerd-dictation at github that describes the same problem like my problem. It is the issue number 120.

**Comment by ralf3u** [comment]
This is what I did in the past:

**installation of vosk**
```
pip3 install vosk
sudo apt install python3-pip
pip3 install vosk
```
restart the computer

**download model**
`alphacephei.com/vosk/models`

**installation of nerd-dictation**
`git clone https://github.com/ideasman42/nerd-dictation.git`

**installation of further sotware**
sudo apt install xdotool

**Comment by ralf3u** [comment]
**Usage of Vosk**
1. record speech for example in Audacity
2. save the speech, for example 1.ogg, in the USER-File, so /home/USER/
3.  in the terminal
`vosk-transcriber -m FULL-PATH-OF-MODEL -i 1.ogg -o 1.txt`
example:
`vosk-transcriber -m /home/USER/model-en/ -i 1.ogg -o 1.txt`

Result: The text of the speech is in 1.txt.

**Comment by ralf3u** [comment]
**Usage of the punctuation** (if `sudo apt install xdotool` can be excecuted)

This is how I used that in the past:

1. paste the text-document, like 1.txt, inside the folder of the model for the punctuation
2. terminal: `python3 example.py 1.txt > 2.txt`

Result: In 2.txt there is the text with punctuation.

**Comment by ralf3u** [comment]
**Usage of nerd-dictation**

This is how I used it in the past:
nerd-dictation should be automatically in the USER-file after the installation

in the terminal:
`cd nerd-dictation/ && python3 nerd-dictation begin --vosk-model-dir=PATH`
example:
`cd nerd-dictation/ && python3 nerd-dictation begin --vosk-model-dir=/home/USER/model-en/`
change to Geany and speak
to stop the speech-to-text recognition: change to the terminal and press Ctrl+c

**Comment by ralf3u** [comment]
**Usage of shortcuts to start and stop dictation in nerd-dictation**

Command to start for openbox:
`bash -c 'cd nerd-dictation && python3 nerd-dictation begin --vosk-model-dir=/home/USER/model-en'`
Command to stop for openbox:
`bash -c 'cd nerd-dictation && python3 nerd-dictation end'`

In rc.xml:
```
    <keybind key="W-+">
      <action name="Execute">
        <command>paste command from above to start</command>
      </action>
    </keybind>
    <keybind key="W-#">
      <action name="Execute">
        <command>paste command from above to stop</command>
      </action>
    </keybind>
```

**Comment by ralf3u** [comment]
It seems to be that the maintainer of nerd-dictation stopped the development of nerd-dictation, but maybe I'm wrong. If you find a way to make it run, what about the idea to fork nerd-dictation and call it xiota-dictation?

**Comment by ralf3u** [comment]
Did you try already Vosk? If the answer is yes, do you like it? Did you find a way to make nerd-dictation run?

**Comment by xiota** [comment]
Haven't tried yet.  Plan to eventually.  I did see that [github.com/ideasman42/nerd-dictation](https://github.com/ideasman42/nerd-dictation) was updated a month ago.
